06_gpu_and_ml/yolo/yolo.py

```diff
@@ -0,0 +1,288 @@
+# which has been trained to perform bounding box object detection.
+#
+# In this example, we use the V10 model via the [Ultralytics package](https://docs.ultralytics.com/). We will:
+# - Fine tune on a custom dataset, in our case the DoclayNet dataset for object detection within documents.
+# - Deploy an inference endpoint.
+
+# For commercial use, sure to consult the [ultralytics software license](https://docs.ultralytics.com/#yolo-licenses-how-is-ultralytics-yolo-licensed).
+
+# ## Building Our Image
+#
+# The example uses the Ultralytics package to run training inference and the Huggingface dataset package to load DoclayNet data.
+# 
+# All packages are installed into a Pytorch 2.3.1 base image
+# using the `pip_install` function.
+#
+# In addition, we download the base YOLO model weights into the container at build-time
+# using the `run_function` function
+
+import modal
+import urllib.request, os, yaml, json, shutil, zipfile
+from pathlib import Path
+
+def load_yolo_weights():
+    from ultralytics import YOLO
+    # Choose the large size checkpoint
+    # Other model sizes can be found at: https://docs.ultralytics.com/models/yolov10/#performance
+    model = YOLO('yolov10l.pt') # This downloads the model's base weights if not already present in the image.
+    return model
+
+yolo_image = (
+    modal.Image.from_registry("pytorch/pytorch:2.3.1-cuda11.8-cudnn8-runtime")
+    .apt_install(["libgl1-mesa-glx", "libglib2.0-0"])
+    .pip_install(["ultralytics", "datasets", "opencv-python"])
+    .run_function(load_yolo_weights)
+
+app = modal.App("train-yolo")
+
+# ## Downloading The Dataset
+# 
+# Similar to downloading model weights at build-time, it may be tempting to also download the DoclayNet dataset at build-time.
+# We instead choose to download the dataset into a Modal Volume, using a Modal Function.
+# This approach has two benefits:
+# - Reduces the size of the container image, making cold boots much faster. This is especially important for inference when there's no need for having the data in the container.
+# - Makes it easier to train on many datasets, storing to create many finetunes usable for inference.
+
+# Define a Volume, which will persist training data and trained weights between invocations 
+volume = modal.Volume.from_name("doclaynet-base-yolo", create_if_missing=True)
+
+def download_file(url, output_path):
+    from tqdm import tqdm
+    class DownloadProgressBar(tqdm):
+        def update_to(self, b=1, bsize=1, tsize=None):
+            if tsize is not None:
+                self.total = tsize
+            self.update(b * bsize - self.n)
+    with DownloadProgressBar(unit='B', unit_scale=True, miniters=1, desc=url.split('/')[-1]) as t:
+        urllib.request.urlretrieve(url, filename=output_path, reporthook=t.update_to)
+
+def convert_dataset_to_yolo(dataset_path, output_path):
```

Comment:
Surprised at the length of this code -- is there not a utility we can pull from ultralytics?